packages/amplify-provider-awscloudformation/src/utils/consolidate-apigw-policies.ts

```diff
@@ -0,0 +1,239 @@
+import * as path from 'path';
+import { $TSAny, $TSContext, $TSObject, FeatureFlags, JSONUtilities } from 'amplify-cli-core';
+import * as iam from '@aws-cdk/aws-iam';
+import * as cdk from '@aws-cdk/core';
+import { prepareApp } from '@aws-cdk/core/lib/private/prepare-app';
+import { S3 } from '../aws-utils/aws-s3';
+import { ProviderName } from '../constants';
+import { getResourceDirPath } from '../resourceParams';
+
+type ApiGatewayAuthStackProps = Readonly<{
+  description: string;
+  stackName: string;
+  apiGateways: any[];
+}>;
+
+type ApiGatewayPolicyCreationState = {
+  apiGateway: $TSAny;
+  apiRef: cdk.CfnParameter;
+  env: cdk.CfnParameter;
+  path: $TSAny;
+  methods: $TSAny;
+  roleCount: number;
+  roleName: cdk.CfnParameter;
+  policyDocSize: number;
+  managedPolicy: iam.CfnManagedPolicy;
+  namePrefix: string;
+};
+
+export const APIGW_AUTH_STACK_LOGICAL_ID = 'APIGatewayAuthStack';
+const API_PARAMS_FILE = 'api-params.json';
+const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
+const MAX_MANAGED_POLICY_SIZE = 6_144;
+const S3_UPLOAD_PATH = 'api/apiGwAuthStackTemplate.json';
+
+export class ApiGatewayAuthStack extends cdk.Stack {
+  constructor(scope: cdk.Construct, id: string, props: ApiGatewayAuthStackProps) {
+    super(scope, id, props);
+    this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
+
+    const authRoleName = new cdk.CfnParameter(this, 'authRoleName', {
+      type: 'String',
+    });
+    const unauthRoleName = new cdk.CfnParameter(this, 'unauthRoleName', {
+      type: 'String',
+    });
+    const env = new cdk.CfnParameter(this, 'env', {
+      type: 'String',
+    });
+
+    new cdk.CfnCondition(this, 'ShouldNotCreateEnvResources', {
+      expression: cdk.Fn.conditionEquals(env, 'NONE'),
+    });
+
+    let authRoleCount = 0;
+    let unauthRoleCount = 0;
+    let authPolicyDocSize = 0;
+    let unauthPolicyDocSize = 0;
+    let authManagedPolicy;
+    let unauthManagedPolicy;
+
+    props.apiGateways.forEach(apiGateway => {
+      if (!Array.isArray(apiGateway.params.paths)) {
+        return;
+      }
+
+      const apiRef = new cdk.CfnParameter(this, apiGateway.resourceName, {
+        type: 'String',
+      });
+
+      const state: ApiGatewayPolicyCreationState = {
```

Comment:
why does this need to be outside of the forEach closure below?

Reply:
It doesn't strictly need to be. I hoisted it out of that loop to avoid creating a bunch of extra garbage objects. It could be hoisted up even further, but my thought was that the number of API gateways is probably small, but the number of paths could be a good bit larger.